dali/python/nvidia/dali/external_source.py

```diff
@@ -62,31 +62,51 @@ def __next__(self):
             return next(self.it)
 
 class _ExternalSourceGroup(object):
-    def __init__(self, callback, is_multioutput, instances = [], cuda_stream = None, use_copy_kernel = None):
+    def __init__(self, callback, is_multioutput, instances = [], cuda_stream = None, use_copy_kernel = None, batch = True):
         self.instances = list(instances)  # we need a copy!
         self.is_multioutput = is_multioutput
         self.callback = callback
         self._cuda_stream = cuda_stream
         self.use_copy_kernel = use_copy_kernel
+        self.batch = batch
         if callback is not None:
-            if callback.__code__.co_argcount not in [0, 1]:
-                raise TypeError("External source callback must be a callable with 0 or 1 argument")
-            self.accepts_iter_num = (callback.__code__.co_argcount == 1)
+            if batch:
+                if callback.__code__.co_argcount not in [0, 1]:
+                    raise TypeError("External source callback must be a callable with 0 or 1 argument")
+            else:
+                if callback.__code__.co_argcount not in [0, 1, 2]:
+                    raise TypeError("Per-sample external source callback must be a callable with up to 2 arguments: " +
+                                    "``sample_index``, ``iter_index``")
+            self.accepts_iter_num = (callback.__code__.co_argcount == (1 if batch else 2))
+            self.accepts_sample_index = (not batch and callback.__code__.co_argcount > 0)
         else:
-            self.accepts_iter_num = None
+            self.accepts_iter_num = False
+            self.accepts_sample_index = False
 
     def append(self, instance):
         self.instances.append(instance)
 
-    def call_and_feed(self, pipeline, current_iter):
+    def callback_args(self, sample, iter):
+        args = []
+        if self.accepts_sample_index:
+            args.append(sample)
         if self.accepts_iter_num:
-            callback_out = self.callback(current_iter)
+            args.append(iter)
+        return args
+
+    def call_and_feed(self, pipeline, batch_size, current_iter):
+        if self.batch:
+            callback_out = self.callback(*self.callback_args(None, current_iter))
         else:
-            callback_out = self.callback()
+            callback_out = [self.callback(*self.callback_args(i, current_iter)) for i in range(batch_size)]
 
         if self.is_multioutput:
             for op in self.instances:
-                data = callback_out[op._output_index]
+                if self.batch:
+                    data = callback_out[op._output_index]
+                else:
+                    # extract a single output
+                    data = [callback_out[i][op._output_index] for i in range(batch_size)]
```

Comment:
I am not sure if it matters later on in the pipeline but maybe user should be able to pass optional collate function to decide how exactly the list of multioutput samples should be collated here? E.g. to return [array(128x128x3)], matrxi(batch_size x 1) rather than [array(128x128x3)], [array(1)]? Or lists are just fine here?

Reply:
I don't quite follow.
The values from multi-output external source are returned as:
```(batch1, batch2, ...)``` for batched variant and as 
```(sample_from_batch1, sample_from_batch2, ...)``` for per-sample variant.
One of the reason why the batched variant is expected to produce a tuple of batches (as opposed to a batch of tuples) is that the batches can be contiguous tensors where the outermost dimension is batch:
```(batch_tensor, [sample_tensor, sample_tensor, sample_tensor])```

Reply:
Okok, so in the per-sample variant in case of two outputs, we'll be now necessarily returning tuple of (two) lists (each with length of batch size), right? And what I've been wondering is if we should allow user to pass here some additional callback that would for example squeeze one of this lists into batch tensor so that the output would be (batch_tensor, [sample1_tensor, sample2_tensor, sample3_tensor...]).

Reply:
Basically I am referring to something like collate_fn from pytorch https://pytorch.org/docs/stable/data.html that given callback_out that is [sample_from_batch1, sample_from_batch2, ...] leaves it to the user how exactly it should be transformed into tuple of batches.

Reply:
> Okok, so in the per-sample variant in case of two outputs, we'll be now necessarily returning tuple of (two) lists (each with length of batch size)
In a two-output, per-sample case the callback returns a tuple of two _samples_ (e.g. an image and a label).
In a two-output, per-batch case, the callback returns two _batches_ (e.g. a batch of images and a batch of labels).

> Basically I am referring to something like collate_fn 
I think it's not necessary at this point - when you want some specific way of assembling the batch, you can just feed the batches directly. It _might_ help in some cases when we have parallel processing, but I think that would be shifting the functionality of DALI pipeline into external source. We can discuss that later.

Reply:
Okok!